### PR TITLE
GH-36692: [CI][Packaging] Pin gemfury to 0.12.0 due to issue with faraday dependency

### LINKED ISSUE #36692: [CI][Packaging] gemfury client fails to update wheels on ARM64 and Windows
### Describe the bug, including details regarding any error messages, version, and platform.

The latest nightlies wheels have failed for:
[wheel-manylinux-2-28-cp310-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552152433/jobs/10139129003)
[wheel-manylinux-2-28-cp311-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552158602/jobs/10139143530)
[wheel-manylinux-2-28-cp38-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552161248/jobs/10139150055)
[wheel-manylinux-2-28-cp39-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552148318/jobs/10139120029)
[wheel-manylinux-2014-cp310-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552154023/jobs/10139132957)
[wheel-manylinux-2014-cp311-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552148717/jobs/10139121077)
[wheel-manylinux-2014-cp38-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552153669/jobs/10139131956)
[wheel-manylinux-2014-cp39-arm64](https://github.com/ursacomputing/crossbow/actions/runs/5552155799/jobs/10139137617)
[wheel-windows-cp310-amd64](https://github.com/ursacomputing/crossbow/actions/runs/5552160772/jobs/10139148593)
[wheel-windows-cp311-amd64](https://github.com/ursacomputing/crossbow/actions/runs/5552153112/jobs/10139130602)
[wheel-windows-cp38-amd64](https://github.com/ursacomputing/crossbow/actions/runs/5552162078/jobs/10139152473)
[wheel-windows-cp39-amd64](https://github.com/ursacomputing/crossbow/actions/runs/5552154671/jobs/10139134695)

For windows the error is:
```
ERROR:  Error installing gemfury:
	The last version of faraday (< 2.0.0.pre, >= 0.9.0) to support your Ruby & RubyGems was 1.10.3. Try installing it with `gem install faraday -v 1.10.3` and then running the current command again
	faraday requires Ruby version >= 2.6. The current ruby version is 2.5.0.
Successfully installed progressbar-1.13.0
Successfully installed highline-2.0.3
Successfully installed ruby2_keywords-0.0.5
Successfully installed multipart-post-2.3.0
Error: Process completed with exit code 1.
```
For ARM:
```
/home/runner/.local/share/gem/ruby/3.0.0/gems/gemfury-0.12.1/lib/faraday/adapter/fury_http.rb:5:in `<class:Adapter>': uninitialized constant Faraday::Adapter::NetHttp (NameError)
	from /home/runner/.local/share/gem/ruby/3.0.0/gems/gemfury-0.12.1/lib/faraday/adapter/fury_http.rb:4:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /home/runner/.local/share/gem/ruby/3.0.0/gems/gemfury-0.12.1/lib/gemfury.rb:15:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /home/runner/.local/share/gem/ruby/3.0.0/gems/gemfury-0.12.1/bin/fury:7:in `<top (required)>'
	from /home/runner/.local/share/gem/ruby/3.0.0/bin/fury:25:in `load'
	from /home/runner/.local/share/gem/ruby/3.0.0/bin/fury:25:in `<main>'
Error: Process completed with exit code 1.
```

There seems to be an issue with the `faraday` library.

### Component(s)

Continuous Integration, Packaging

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

Some nightly wheel jobs have failed to upload.

### What changes are included in this PR?

Install required gem dependency.

### Are these changes tested?

Yes, crossbow tasks

### Are there any user-facing changes?

No
* Closes: #36692

### CODE REVIEW

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
I've just realised the previous jobs won't fail because we check the branch is the default branch. I'll temporarily remove the check and resubmit the jobs again.

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
I am confused because this PR was triggered yesterday without this change and the wheels jobs worked: https://github.com/apache/arrow/pull/36657

**Comment by kou** [comment]
It seems that gemfury 0.12.1 was released yesterday :-)
https://rubygems.org/gems/gemfury/versions/0.12.1

**Comment by kou** [comment]
Another approach (ensure using recent Ruby):

```diff
diff --git a/dev/tasks/macros.jinja b/dev/tasks/macros.jinja
index 4c65a64c8..a879dcb67 100644
--- a/dev/tasks/macros.jinja
+++ b/dev/tasks/macros.jinja
@@ -137,6 +137,11 @@ on:
     run: |
       sudo apt update
       sudo apt install -y ruby-full
+  - name: Set up Ruby by GitHub Actions
+    if: !(runner.os == 'Linux' && runner.arch != 'X64')
+    use: ruby/setup-ruby@1
+    with:
+      ruby-version: "ruby"
   - name: Upload package to Gemfury
     shell: bash
     run: |
```

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

File: dev/tasks/macros.jinja [hunk 1]
```diff
@@ -131,12 +131,16 @@ on:
...
 {%- macro github_upload_gemfury(pattern) -%}
-  {%- if arrow.is_default_branch() -%}
   - name: Set up Ruby by apt
     if: runner.os == 'Linux' && runner.arch != 'X64'
     run: |
       sudo apt update
       sudo apt install -y ruby-full
+  - name: Set up Ruby by GitHub Actions
+    if: !(runner.os == 'Linux' && runner.arch != 'X64')
+    use: ruby/setup-ruby@1
```
Comment (kou) [inline dev/tasks/macros.jinja:141]:
```suggestion
    uses: ruby/setup-ruby@1
```

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

File: dev/tasks/macros.jinja [hunk 2]
```diff
@@ -131,12 +131,16 @@ on:
...
 
 {%- macro github_upload_gemfury(pattern) -%}
-  {%- if arrow.is_default_branch() -%}
   - name: Set up Ruby by apt
     if: runner.os == 'Linux' && runner.arch != 'X64'
     run: |
       sudo apt update
       sudo apt install -y ruby-full
+  - name: Set up Ruby by GitHub Actions
+    if: !(runner.os == 'Linux' && runner.arch != 'X64')
```
Comment (kou) [inline dev/tasks/macros.jinja:140]:
```suggestion
    if: |
      !(runner.os == 'Linux' && runner.arch != 'X64')
```

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

File: dev/tasks/macros.jinja [hunk 3]
```diff
@@ -131,12 +131,17 @@ on:
...
-  {%- if arrow.is_default_branch() -%}
   - name: Set up Ruby by apt
     if: runner.os == 'Linux' && runner.arch != 'X64'
     run: |
       sudo apt update
       sudo apt install -y ruby-full
+  - name: Set up Ruby by GitHub Actions
+    if: |
+      !(runner.os == 'Linux' && runner.arch != 'X64')
+    uses: ruby/setup-ruby@1
```
Comment (kou) [inline dev/tasks/macros.jinja:142]:
```suggestion
    uses: ruby/setup-ruby@v1
```

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

File: dev/tasks/macros.jinja [hunk 4]
```diff
@@ -131,12 +131,17 @@ on:
...
 {%- macro github_upload_gemfury(pattern) -%}
-  {%- if arrow.is_default_branch() -%}
   - name: Set up Ruby by apt
     if: runner.os == 'Linux' && runner.arch != 'X64'
     run: |
       sudo apt update
       sudo apt install -y ruby-full
+  - name: Set up Ruby by GitHub Actions
+    if: |
+      !(runner.os == 'Linux' && runner.arch != 'X64')
```
Comment (kou) [inline dev/tasks/macros.jinja:141]:
```suggestion
    if: runner.arch == 'X64' && runner.os != 'macOS'
```

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

**Comment by kou** [comment]
Hmm. It seems that gemfury 0.12.1's dependency ([`faraday >= 0.9.0, < 2.0.0.pre`) is buggy...
It accepts `faraday-2.0.0.alpha.pre.4`...

We may need to pin to gemfury 0.12.0 for now...: `gem install --user-install gemfury -v 0.12.0`

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-windows-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-windows-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp310-arm64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp39-arm64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2014-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-windows-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp311-arm64

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@kou this should be good to review now. All wheels have been uploaded successfully.

File: dev/tasks/macros.jinja [hunk 5]
```diff
@@ -137,11 +137,25 @@ on:
...
+  - name: Set up Ruby by GitHub Actions
+    if: runner.arch == 'X64' && runner.os != 'macOS'
+    uses: ruby/setup-ruby@v1
+    with:
+      ruby-version: "ruby"
+  - name: Install gemfury client on ARM self-hosted
+    if: runner.arch != 'X64'
+    run: |
+      # GH-36692: Pin gemfury due to wrong faraday dependency declaration.
+      gem install --user-install gemfury -v 0.12.0
```
Comment (assignUser) [inline dev/tasks/macros.jinja:149]: Is there a downside to  the user install option? If not should we just install that way on all runners?
Alternatively these steps could be merged and the option inserted like this `{{ "${{ runner.arch == 'X64' && '--user-install' || ''}}" }}`

File: dev/tasks/macros.jinja [hunk 6]
```diff
@@ -137,11 +137,25 @@ on:
...
+    if: runner.arch == 'X64'
+    run: |
+      # GH-36692: Pin gemfury due to wrong faraday dependency declaration.
+      gem install gemfury -v 0.12.0
   - name: Upload package to Gemfury
     shell: bash
     run: |
-      PATH=$(echo $(ruby -r rubygems -e 'puts Gem.user_dir') | sed "s/C:\//\/c\//")/bin:$PATH
-      gem install --user-install gemfury
+      PATH=$(echo $(ruby -r rubygems -e 'puts Gem.user_dir'))/bin:$PATH
```
Comment (kou) [inline dev/tasks/macros.jinja:158]: Could you move this to `name: Install gemfury client on ARM self-hosted` by replacing with `ruby -r rubygems -e 'puts("#{Gem.user_dir}/bin")' >> $GITHUB_PATH`?

**Review by kou — APPROVED** [review]
+1

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

File: dev/tasks/macros.jinja (hunk 5 above)
Comment (raulcd) [reply dev/tasks/macros.jinja:148]: If we want to use the `setup-ruby` action, to ensure recent ruby versions, the step fails when trying to use `--user-install` that's why we are separating them. As suggested by Kou I've also added adding the user install dir to the path on the ARM step.
I also prefer to be more explicit on the difference between ARM and non-ARM on the different step. I feel is more readable than:
`{{ "${{ runner.arch == 'X64' && '--user-install' || ''}}" }}`

File: dev/tasks/macros.jinja (hunk 6 above)
Comment (raulcd) [reply dev/tasks/macros.jinja]: Thanks @kou , I realized the PATH extension wasn't reused on the different steps. I did not think about `GITHUB_PATH`. I am trying it out now.

**Pull request merged**